rexfw/statistics/writers/graphite.py

```diff
@@ -0,0 +1,74 @@
+from pickle import dumps
+from struct import pack
+from socket import socket
+
+# timestamp is the number of seconds since unix epoch time.
+# Carbon-cache will use the time of arrival if the timestamp is set to -1.
+# https://graphite.readthedocs.io/en/stable/feeding-carbon.html
+
+
+def _make_tuples_for_stepsize(job_name, elements):
+    listOfStatsTuples = []
+    elements = [e for e in elements if e.name == "stepsize"]
+    for e in elements:
+        timestamp = -1
+        listOfStatsTuples.append((job_name + ".stepsize", (timestamp, e.current_value)))
+    return listOfStatsTuples
+
+
+def _make_tuples_for_acceptance_rates(job_name, elements):
+    listOfStatsTuples = []
+    elements = [e for e in elements if e.name == "acceptance rate"]
+    for e in elements:
+        timestamp = -1
+        if len(e.origins) == 1:
+            listOfStatsTuples.append(
+                (
+                    job_name + ".acceptance_rate.local",
+                    (timestamp, e.current_value),
+                )
+            )
+        if len(e.origins) == 2:
+            listOfStatsTuples.append(
+                (
+                    job_name + ".acceptance_rate.re",
+                    (timestamp, e.current_value),
+                )
+            )
+    return listOfStatsTuples
+
+
+class GraphiteStatisticsWriter:
+    """
+    Writes statistics to Graphite server.
+    :param str job_name
+    :param str graphite_url
+    :param int graphite_port
+    """
+
+    def __init__(self, job_name, graphite_url="127.0.0.1", graphite_port=2004):
+
+        self.job_name = job_name
+        self.sock = socket()
+        self.sock.connect((graphite_url, graphite_port))
+
+    def _write(self, elements, make_tuples):
+        # make_tuples should export list of (path, (timestamp, value))
+        # for more info look at
+        # https://graphite.readthedocs.io/en/stable/feeding-carbon.html
+        listOfStatsTuples = make_tuples(self.job_name, elements)
+
+        payload = dumps(listOfStatsTuples, protocol=2)
+        size = pack("!L", len(payload))
+        self.sock.sendall(size)
+        self.sock.sendall(payload)
+        # TODO: for large size of data, they should be pushed once a certain number of samples
+        # (METRICS_BUFFER_SIZE) are buffered or upon a defined period of time (MAX_BUFFER _TIME)
+        # For more info look at
+        # https://www.metricfire.com/blog/monitoring-with-graphite-installation-and-setup/
+
+    def write_stepsize(self, elements):
```

Comment:
It would be great if the `write()` methods maintained the original interface of `AbstractStatisticsWriter`, namely that they also take the `step` argument. Do you think we could use the `step` (which is an integer telling us the number of MCMC samples that have been drawn) as a timestamp instead of an actual time stamp?